### DISCUSSION TITLE
Using NextJS in optimized way for some use case

### DISCUSSION THREAD

**Discussion by ScreamZ** [vercel/next.js#67896] [Help]
### Summary

Hello,

I'm currently using a NextJS 14 app currently on vercel but maybe on my barebone server soon, and I have an issue with the Next image for a specific use case.

I've made detailed and a repro here https://stackoverflow.com/questions/78763699/reusing-react-component-across-the-three-and-deduplicate-request-nextjs-or-rea

Maybe you could help me, I'm consuming much bandwidth without knowing why exactly and what can I do, while keeping NextImage component.

Thanks

### Example

https://codesandbox.io/p/sandbox/blissful-hill-qx9v5

@leerob I'm pretty sure you're an expert on that and this could make a wonderful video on youtube channel ;)

**Comment by icyJoseph** [comment]
I think you have a misunderstanding of what `<Component />` really is...

In React, we don't do `Component()`, rather we do `<Component />`. And the difference matters a lot!


Doing `<Component />` is called, creating a React Element. React Components, are functions that return React Elements. And React Elements, are an object description of UI, which we represent using `JSX`, but really, are just plain objects.

So when you have:

```jsx
<h1 className="prose">Hello World</h1>
```

What really exists, is:

```js
{
  type: "h1",
  props: {
    children: "Hello World",
    className: "prose"
  },
 // and other data
}
```

At the end of the day, formally, we used to say, that JSX are calls to React.createElement.

React uses that object description to know what to do, render a `div` in `React.createElement("div", /* other stuff */)`, or call a function that returns more information in `React.createElement(MyComponent, /* other stuff */)`. Nowadays, this is not 100% accurate, but it is still a valid model.

Alright, so in your case, you want the component that renders the image, to happen once, only. So, you could break the rules, and do this:

```tsx
let id = 0;

const CustomNode = () => {
  console.log("Render node", ++id);

  return (
    // Added some fake params, to make the browser not use it's cache.
    <img
      src="https://via.placeholder.com/150?fakingsomenextImageParams=cool"
      alt="placeholder"
      width={200}
      height={200}
    />
  );
};

// const reusedNode = <CustomNode />;

export const Repro: React.FC = () => {
  const imgNode = useMemo(() => CustomNode() /* I really dislike this */, []);
  const [list, setList] = useState([imgNode]);

  return (
    <div>
      <button onClick={() => setList((prev) => [...prev, imgNode])}>Add</button>
      {list.map((node, index) => (
        <div key={index}>
          {node} {index}
        </div>
      ))}
    </div>
  );
};
```

In StrictMode, that'll still run twice though, but no more. 

This basically, collects the UI description, and stores it, manually. It is very fragile!

I think this is a bad, and poor, solution though. A render pass shouldn't be causing side effects like this. Neither should we necessarily, rely on how many times a render pass happens. These can happen several times! Also, rendering is just a way to collect UI description... something else is wrong here...

- https://kentcdodds.com/blog/dont-call-a-react-function-component

If I read correctly, you want to make the request for that image only once, but for some reason, the image has query parameters, that prevent the browser from caching it, causing several unnecessary requests, that end up returning the same asset anyway.

How is it that you end up in that situation? Could you show us more? Couldn't you somehow, try to normalize, the image url, or drop element from the url, so it ends up being the same request? As it should be?

**Reply by ScreamZ** [reply]
Thank you for your kind explanation.

The render id++ is just an example to represent that the component is re-rendering. I don't have this in production or in development.

As I'm using a React Server Component, I don't need to useMemo (as the serialized payload act the same) (correct me if I'm wrong, but even by trying, it achieve the same output as you give with useMemo).

The issue that remains is the URL get parameter caused by this component optimization : https://nextjs.org/docs/app/api-reference/components/image

That gives URL like this one : http://localhost:3000/_next/image?url=https%3A%2F%2Fcdn.discordapp.com%2Favatars%2F165191437376290817%2F315d2ac86db1f78c2b37251aa104576d.png&w=256&q=75

For optimization purposes.

So in my case, and specifically for it, as the image is replicated. I should rather rely on standard `<img>` tag and try to give a fixed URL without parameters.

**Reply by icyJoseph** [reply]
Or you could pass the `unoptimized` flag to the `next/image` component

**Reply by ScreamZ** [reply]
Yep, just did it. Even though still experiencing the multiple load.

My exact code is as follow:

React Server component
```tsx
 <ChatContextProvider
        initialEmotion={initialEmotion}
        // As its server side renderer and serialized to client, its already memoized
        users={{
          assistant: <ChatUser role="assistant" name={char.name} image={char.image} />,
          user: (
            <ChatUser
              role="user"
              name={session.user.displayName ?? t("unknown")}
              image={session.user.image ?? undefined}
            />
          ),
        }}
      >
        <ChatEngine
          initialCursor={initialMessages[0].id}
          initialConversation={initialMessages.map((message) => (
            <ChatMessageUI key={message.id} role={message.role} content={message.content} />
          ))}
          userUnlockedAchievementsForSupa={userUnlockedAchievementsForSupa}
          conversation={conversation}
          char={char}
        />
      </ChatContextProvider>
```

```tsx
"use client";
import { createContext, use, useCallback, useMemo, useState } from "react";
import { z } from "zod";

export type ChatContextType = {
  users: {
    assistant: React.ReactNode;
    user: React.ReactNode;
  };
  activeEmotion: string;
  setActiveEmotion: (emotion: string) => void;
};

const ChatContext = createContext<ChatContextType>({
  users: {
    assistant: null,
    user: null,
  },
  activeEmotion: "neutral",
  setActiveEmotion: () => {
    console.warn("No chat context provided.");
  },
});

type ChatContextProviderProps = {
  initialEmotion?: SupaEmotion;
  users: ChatContextType["users"];
  children: React.ReactNode;
};

export const ChatContextProvider = ({
  children,
  users,
  initialEmotion = "neutral"}: ChatContextProviderProps) => {
  const [activeEmotion, setActiveEmotion] = useState(initialEmotion);

  const value = useMemo<ChatContextType>(
    () => ({
      activeEmotion,
      setActiveEmotion: (emotion: string) => {
        setActiveEmotion(z.nativeEnum(["neutral", "happy"]).catch("neutral").parse(emotion));
      },
      users,
    }),
    [activeEmotion, setActiveEmotion, users],
  );

  return <ChatContext.Provider value={value}>{children}</ChatContext.Provider>;
};

export const useChatContext = () => use(ChatContext);
```

```tsx
export const ChatMessageUI: React.FC<ChatMessageUIProps> = (props) => {

  const chatContext = useChatContext();

  return (
    <div className="flex flex-col gap-2">
      {props.role !== "system" && (
        <div className={props.role === "user" ? "ml-auto" : undefined}>{chatContext.users[props.role]}</div>
      )}
    </div>
  );
};
```

Each time a messageUI is rendered and streamed from server. Maybe I'm losing the reference due to streaming. But as the context is supposed to be client side only, this should be a safe reference
